Specification.md

```diff
@@ -0,0 +1,87 @@
+# XApp Specification
+
+TBD: I suppose it would be nice to call the project XApp instead of xapp ? (used lower case for now)
+
+In order to be shown on the xapp project, applications need to fulfill a list of requirements. This specification is used in order to summarize these requirements and as such can be utilized to decide if an application can join the xapp project.
+
+## Requirements
+
+### Licensing
+
+Xapps need to be licensed on a [FOSS license](https://opensource.org/licenses).
+
+### Operating System
+
+Xapps need to be able to run in a Linux environment. The support of other operating systems is optional.
+
+### Graphical Desktop Application
+
+This project targets graphical applications which run on desktop systems and dependencies of them. E.g. console applications, kernel-modules and system-services cannot be xapps.
+
+### Usage of standards
+
+In order to be desktop-environment agnostic, the compliance to the following [freedesktop.org](https://www.freedesktop.org) standards is mandatory to be listed as an xapp:
+
+- [Xdg Autostart](https://www.freedesktop.org/wiki/Specifications/autostart-spec/)
+- [Xdg Base Directories](https://www.freedesktop.org/wiki/Specifications/basedir-spec/)
+- [Xdg Desktop Entry](https://www.freedesktop.org/wiki/Specifications/desktop-entry-spec/)
+- [Xdg Desktop Menus](https://www.freedesktop.org/wiki/Specifications/menu-spec/)
+- [File Manager DBUS interface](https://www.freedesktop.org/wiki/Specifications/file-manager-interface/)
+- [File URI Specification](https://www.freedesktop.org/wiki/Specifications/file-uri-spec/)
+- [Free Media Player Specification](https://www.freedesktop.org/wiki/Specifications/free-media-player-specs/)
+- [Free Media Player Remote Interface Specification](https://www.freedesktop.org/wiki/Specifications/mpris-spec/)
+- [Xdg Icon Theme](https://www.freedesktop.org/wiki/Specifications/icon-theme-spec/)
+- [Xdg Icon Naming](https://www.freedesktop.org/wiki/Specifications/icon-naming-spec/)
+- [Shared mime Database](https://www.freedesktop.org/wiki/Specifications/shared-mime-info-spec/)
+- [Startup Notification](https://www.freedesktop.org/wiki/Specifications/startup-notification-spec/)
+- [Trashcan support](https://www.freedesktop.org/wiki/Specifications/trash-spec/)
+- [Bookmark Interchange Format](https://pyxml.sourceforge.net/topics/xbel/)
+
+TBD: Not sure on these:
+- [MIME Applications Associations](https://www.freedesktop.org/wiki/Specifications/mime-apps-spec/)
+- [XSettings](https://www.freedesktop.org/wiki/Specifications/xsettings-spec/)
+- [Cursor Conventions](https://www.freedesktop.org/wiki/Specifications/cursor-spec/)
+- [Sound Theme](https://www.freedesktop.org/wiki/Specifications/sound-theme-spec/)
+
+### Window decorations
+
+Applications which are part of the xapp project should support server side window decorations and use them by default. Xapps are free to optionally as well support client side decorations [^1].
```

Comment:
I don't see how this is enforceable. GNOME Wayland notoriously doesn't support server-side decorations and instead applications have to resort to using things like libdecor to generate client-side decorations.

Reply:
Xapps which want to be able to run in GNOME wayland can still use CSD. Though, the default should be SSD.

An app which wants to support GNOME wayland might want to probe if it is running in a GNOME wayland session at startup, and if so, it can use CSD.

Reply:
Imo the fact that gnome does not support SSD can be considered as a GNOME problem ... personally I don't plan to install optional CSD into Xfce apps, just to please gnome ideas.

Reply:
For Wayland, apps should be responsive to [KDE server decoration](https://wayland.app/protocols/kde-server-decoration)/[XDG decoration](https://wayland.app/protocols/xdg-decoration-unstable-v1), though xapps should request/set SSD mode.  If the compositor insists on CSD (or doesn't support either of these protocols), the app will either have to draw CSDs, or just be fine with the idea that they'll be undecorated on compositors with no SSD support (which is fine).  So I think the wording I'd like to see here is more like:

"Applications which are part of the xapp project should default to server-side decorations via a windowing system-specific protocol, if one is available.  If an xapp is run on a desktop environment that prefers or requires applications to draw client-side decorations, the xapp may do so, but is not required by this specification to support this mode of operation."

My rationale here is that I could see a situation where the desktop environment itself has no strong opinion on SSD vs. CSD, but allows the user to choose what to prefer (e.g. Wayfire has such a setting).  The xapp should obey the user's preference , assuming the xapp's developer wants to include CSD support.

Reply:
The wording suggested by @kelnos sounds good and reasonable to me. (As someone who has not that much knowledge on the topic)

@JoshStrobl , would you agree with that suggestion?

Reply:
I think the devs might want to support this in some cases but not others. Let me take an example.

**Scenario 1**

Imagine two calculator projects, both XApps. One has a menubar/SSD, one is CSD. The two teams talk to each others but don't feel the need to merge their efforts. These projects are small anyway. DEs appreciate the choice and don't all use the same calculator.

**Scenario 2**

Something that didn't exist before and is hard to maintain comes up. DEs want to use it but they don't all want the exact same UI.

In this scenario we don't want to duplicate the work into multiple projects, the devs might want to provide configurable UI choices.

------------

All in all, I think it's something that's a case by case thing. It depends largely on who wants to use the app, who's making it, whether or not there are other XApp alternatives etc...

I wouldn't make the SSD a requirement. I think what matters the most is the general intention. If a stakeholder wants to use something, it's in the nature of Xapps to accomodate to a certain degree.

Reply:
Though doesn't an app which only supports CSD collide with the requirement to be freely styleable by the user? IIrc, having user-stylable apps was one of the Xapp targets. Apps which force a specific style on window decorations imo disqualify for that target. (imo same story as styling limitations brought by libadwaita)

If you allow CSD only apps, e.g. Xfce could not stat that "All Xapps will look well integrated in our DE". For the README section `## XApp compliance` we would need to cut out the part `and should integrate well into any desktop environment that has adopted the XApp specification.`, because that won't be the case anymore.

Sure, by allowing SSD-only apps we would have more apps which qualify for beeing an Xapp ... though I think we should rather aim for what we want to accomplish instead of maximizing the number of apps which do fit in.

Reply:
An application can use CSDs and still make them styleable, can they not? I'm not sure that I understand why they have to be mutually exclusive.

Reply:
> An application can use CSDs and still make them styleable, can they not? I'm not sure that I understand why they have to be mutually exclusive.

If the user does select a specific decoration in the window manager (e.g. the location and style of the close/maximize/minimize button ... that's currently possibly via xfwm4), then applications which only offer CSD won't be able to follow the selected window style.

Reply:
That's not entirely true - on GNOME apps use CSD and use the placement settings set by the WM.

Reply:
> That's not entirely true - on GNOME apps use CSD and use the placement settings set by the WM.

Not sure to what you refer with "That".  Sure, GNOME apps do use CSD and are placed by the WM … as far as I can tell, nobody questioned that.

Reply:
> If the user does select a specific decoration in the window manager (**e.g. the location and style of the close/maximize/minimize button** ... that's currently possibly via xfwm4), **then applications which only offer CSD won't be able to follow the selected window style**.

Reply:
> > If the user does select a specific decoration in the window manager (**e.g. the location and style of the close/maximize/minimize button** ... that's currently possibly via xfwm4), **then applications which only offer CSD won't be able to follow the selected window style**.

I still don't get your critics.  Let put an example. I use xfwm4 and choose the decorations to look like this: 

![Bildschirmfoto_2024-07-09_23-32-44](https://github.com/xapp-project/xapp-project/assets/2205293/6c068fb9-5dcc-49a3-89a5-54d733bdc8ef)

That works fine for almost all applications I have installed, because they don't force CSD. But there is e.g. GNOME simple-scan, which does force CSD, and as such is not able to follow the window decoration style I selected:

![Bildschirmfoto_2024-07-09_23-33-03](https://github.com/xapp-project/xapp-project/assets/2205293/cd7d3f9b-c704-4bbb-bb15-7b02e9fe3154)

And because of that inability, it looks out of place when used in Xfce.

It might be nice for a single app-developer, DE or toolkit to decide how decorations will look like, to have an unique look. However, for linux-desktop in general, it is rather bad, since it creates fragmentation and is an annoyance if you want to mix apps cross-desktop.

Reply:
On Wayland (I'd hope?) you can ask the compositor what it wants you to use since many compositors let you choose CSD/SSD globally and sometimes on a per-app basis. CSD tends to decrease accessibility, and I'm sensitive to those issues with Gnome in particular [OT Gnome a11y discrimination rant self-censored 🤬], so if you don't have reason to know the user wants CSD, use SSD is the only reasonable default.

Supporting CSD … Given RH+derivs + Debuntu+derivs pretty much all shipping Gnome default unless the distribution explicitly exists to ship something else … I'd argue "universal" apps probably need to support CSD. If there's no consensus on that, I'd recommend a SHOULD, but I won't protest a MAY too loudly because Gnome is frankly idiotic on this issue … among so many others.

I still use CSD for Firefox because Firefox's UI uses ridiculously too much vertical screen real estate for tabs and browser bar. There just isn't a "compact" theme option. I'd care about that a helluva lot more than I do the pretty themes, but even using the GTK theme doesn't use appreciably less real-estate. Some of that is my font scaling. Some of that is padding around UI elements I really don't care about. It is a great example (even if a singular one) of why a person might prefer CSD in some cases.

Reply:
I don't think the intent is to force users to use SSDs, so perhaps the wording here should be clearer.  What we're trying to get across is: obey the user's preference (either through a hint set by the desktop environment, or through an in-app setting), but supporting CSDs is optional, and lack of CSDs will not prevent an app from being considered an xapp.

Reply:
So perhaps a further revision to my suggested wording:

"Applications which are part of the xapp project should abide by user preferences as to whether server-side or client-side decorations are to be used.  If the windowing system has a protocol for informing apps of the user's preference, it should be honored. However, an xapp is not required to support client-side decorations, and may assume that the windowing system will draw decorations in that case.  An xapp may allow the user to override the windowing system setting, but if it does, by default it must respect the setting advertised by the windowing system."

Reply:
Another possibility, similar, but maybe clearer(?):

"Applications which are part of the xapp project should by default assume that server-side decorations will be drawn.  If the windowing system has a protocol to advertise to applications the user's preference between server- and client-side decorations, the application must honor that preference, with the exception that the application may choose not to support client-side decorations.  Applications may also provide an in-app setting to override this behavior, but the setting must default to honoring the global preference exposed by the windowing system, if any."

Reply:
I thought we were clear that SSD must be supported? My reading of kelnos' wording suggests that to be the case, so I will offer:

"Xapp project applications must support server-side window decorations, where applicable, but may also support client-side decorations. If both are supported, the determination of which to use by default should follow user and environment preferences, falling back to server-side if no default can be determined. Applications may allow users to configure them to use either decoration model."

I feel that wording is still clumsier than I like, but it's the best I can come up with at the moment.